---
title: "class10_structural bioinformatics"
author: "Angie Zhou (PID:A69028746)"
date: 02/09/24
format: pdf
---

## 1. Introduction to the RCSB Protein Data Bank (PDB)

```{r}
Data_file <- "Data Export Summary.csv"
pdbstats= read.csv(Data_file, row.names=1)
summary(pdbstats)
head(pdbstats)
```

```{r}
# code extract function
  commasum <- function(x) {
    # Remomve comma, convert to numeric and sum
    sum(as.numeric(gsub(",", "", x)))
  }
```

```{r}
commasum((pdbstats$X.ray))
```

> **Q1:** What percentage of structures in the PDB are solved by X-Ray and Electron Microscopy. **A1:** X.ray: 84.54 % and EM: 8.72%

```{r}
round(apply(pdbstats, 2, commasum)/commasum(pdbstats$Total)*100, 2)
```

> **Q2:** What proportion of structures in the PDB are protein?

> **A2:** 97.86%

```{r}
protein_structures_total <- pdbstats[1:3,7]
protein_structures_total
round(commasum(protein_structures_total)/commasum(pdbstats$Total)*100, 2)
```

> Q. How does the total number of protein structures iN the PDB relate to the total number of protein sequences in Uniport?

```{r}
186898/250322721*100
```

![snapshot of 1HSG](1HSG.png)

> **Q3:** Type HIV in the PDB website search box on the home page and determine how many HIV-1 protease structures are in the current PDB?

> **A3:** There are 486 HIV-1 protease structures

## 2. Visualizing the HIV-1 protease structure

> **Q4:** Water molecules normally have 3 atoms. Why do we see just one atom per water molecule in this structure?
> **A4:** The resolution is not high enough to see H.

> **Q5:** There is a critical “conserved” water molecule in the binding site. Can you identify this water molecule? What residue number does this water molecule have 
> **A5:**Yes, HOH308

> **Q6:** Generate and save a figure clearly showing the two distinct chains of HIV-protease along with the ligand. You might also consider showing the catalytic residues ASP 25 in each chain and the critical water (we recommend “Ball & Stick” for these side-chains). Add this figure to your Quarto document.

![a nice display showing ASP 25 in the MK1 ligand](1HSG_25D.png)

## 3. Introduction to Bio3D in R

```{r}
library(bio3d)
hiv <- read.pdb("1hsg")
hiv
```

```{r}
attributes(hiv)
aa123(pdbseq(hiv)[25])
```

```{r}
head(hiv$atom)
```

## 4. Reading PDB file data into R

> **Q7:** How many amino acid residues are there in this pdb object?

> **A7:** There are 198 amino acid

> **Q8:** Name one of the two non-protein residues?

> **A8:** HOH/MK1

> **Q9:** How many protein chains are in this structure?

> **A9:** There are 2 protein chains

```{r}
adk <- read.pdb("6s36")
adk
```

normal mode analysis(nma) a bioinformatic method to predict functional motions and large-scale structure changes

```{r}
m <- nma(adk)
plot(m)
```

make a wee movie

```{r}
mktrj(m, file="adk_m7.pdb")
```

Extract a sequence and run a blast

```{r}
s <- pdbseq(adk)
blast <- blast.pdb(s)
plot(blast)
```

get the results from BLAST and download all the top hits.

```{r}
hits <- plot(blast)
hits
```

```{r}
hits$pdb.id
```

```{r}
# Download related PDB files
# files <- get.pdb(hits$pdb.id, path="pdbs", split=TRUE, gzip=TRUE)
```

```{r}
BiocManager::install("msa")
devtools::install_bitbucket("Grantlab/bio3d-view")
```

## 4. Comparative structure analysis of Adenylate Kinase

> **Q10.** Which of the packages above is found only on BioConductor and not CRAN?

> **A10.** The package "msa" is found only on Bioconductor and not on CRAN

> **Q11.** Which of the above packages is not found on BioConductor or CRAN?:

> **A11.** The package "bio3d-view" installed using devtools::install_bitbucket("Grantlab/bio3d-view") is not from Bioconductor or CRAN.

> **Q12.** True or False? Functions from the devtools package can be used to install packages from GitHub and BitBucket?

> **A12.** True

## PCA of all these

```{r}
# Perform PCA
library(bio3d)
aa <- get.seq("1ake_A")
files <- get.pdb(hits$pdb.id, path="pdbs", split=TRUE, gzip=TRUE)
pdbs <- pdbaln(files, fit = TRUE, exefile="msa")
pc.xray <- pca(pdbs)
plot(pc.xray)
```

```{r}
mktrj(pc.xray, file = "pca_movie.pdb")
```

> **Q13.** How many amino acids are in this sequence, i.e. how long is this sequence?

> **A13.** 214 aa
